iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/compaction/IcebergQueryCompactor.java

```diff
@@ -96,16 +106,46 @@ public boolean run(CompactorContext context) throws IOException, HiveException,
         throw new HiveException(ErrorMsg.COMPACTION_NO_PARTITION);
       }
     } else {
-      long partitionHash = IcebergTableUtil.getPartitionHash(icebergTable, partSpec);
+      Pair<Integer, StructProjection> partSpecPair =
+          IcebergTableUtil.getPartitionSpecIdAndStruct(icebergTable, partSpec);
+      int partitionSpecId = partSpecPair.getKey();
+      StructProjection partitionValues = partSpecPair.getValue();
+
+      HiveConf.setBoolVar(conf, ConfVars.HIVE_CONVERT_JOIN, false);
+      conf.setBoolVar(ConfVars.HIVE_VECTORIZATION_ENABLED, false);
       HiveConf.setVar(conf, ConfVars.REWRITE_POLICY, RewritePolicy.PARTITION.name());
       conf.set(IcebergCompactionService.PARTITION_PATH, new Path(partSpec).toString());
 
       Map<String, String> partSpecMap = new LinkedHashMap<>();
       Warehouse.makeSpecFromName(partSpecMap, new Path(partSpec), null);
 
-      compactionQuery = String.format("insert overwrite table %1$s select * from %1$s where %2$s=%3$d " +
-              "and %4$s is not null %5$s %6$s", compactTableName, VirtualColumn.PARTITION_HASH.getName(), partitionHash,
-          VirtualColumn.FILE_PATH.getName(), fileSizePredicate == null ? "" : "and " + fileSizePredicate, orderBy);
+      Types.StructType partitionType = Partitioning.partitionType(icebergTable);
+      String partitionCondition = IntStream.range(0, partitionType.fields().size())
+          .mapToObj(i -> {
+            Types.NestedField field = partitionType.fields().get(i);
+            String name = "`partition`." + HiveUtils.unparseIdentifier(field.name());
+            String type = field.type().toString().equalsIgnoreCase("long") ? "bigint" :
+                field.type().toString();
+
+            Object value = partitionValues.get(i, Object.class);
+            if (type.equalsIgnoreCase("date") && value != null) {
+              value = String.format("date_add('1970-01-01', %s)", value);
```

Comment:
should we use `HiveValueConverter` here?

Reply:
Done